linode/manager
# test: [M3-8951] - Cypress component tests for firewall rules drag and drop `keyboard` interaction

## Description 📝
Add cypress Component tests for firewall rules table drag and drop keyboard interaction which was enhanced in #11127

- These tests confirm that the keyboard interaction for the Firewall Rules table's drag-and-drop feature works as expected.
- The tests also validate that the `CustomKeyboardSensor` functions correctly:
  - We have customized the default `KeyboardSensor` from `dnd-kit` to meet our specific requirements.
  - The sensor we are using is `CustomKeyboardSensor`, rather than the default `KeyboardSensor` from `dnd-kit`.
  - `CustomKeyboardSensor` works as expected on smaller window sizes, especially when the vertical scrollbar is enabled. In contrast, the default `KeyboardSensor` from `dnd-kit` does not function as intended in this scenario.

## Changes  🔄
-  Add Cypress Component tests for firewall rules drag and drop keyboard interaction
-  Refactor, clean-up and consolidate mouse drag-and-drop interactions
-  Add/Improve comments

## Target release date 🗓️
N/A

## How to test 🧪
```
yarn cy:component:run -s "cypress/component/features/firewalls/firewall-rule-table.spec.tsx"
```

OR

Test *firewall-rule-table.spec.tsx* using
```
yarn cy:component
```

- Ensure all tests pass.
- Confirm `CustomKeyboardSensor` works as expected (both for normal window sizes and when a smaller window with a vertical scrollbar is enabled).
   - This can be tested against both `CustomKeyboardSensor` and the `dnd-kit` default `KeyboardSensor`.
     - To test this scenario, replace `CustomKeyboardSensor` with the default `KeyboardSensor` from `dnd-kit` in the `FirewallRuleTable.tsx` component, and vice-versa.
     - All the tests should pass when using `CustomKeyboardSensor` and some of the tests (related to vertical scrollbar) will fail when using the default `KeyboardSensor`, confirming that `CustomKeyboardSensor` is functioning correctly.


<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>


## As an Author, before moving this PR from Draft to Open, I confirmed ✅


</details>
